### PR TITLE
Plane: added ICE_OPTIONS bit for throttle while disarmed

### PR DESCRIPTION
**Pull request by tridge**
This allows for starting and warming up an ICE engine while disarmed

note: needs testing with THR_MIN behaviour

### CODE REVIEW

File: libraries/AP_Arming/AP_Arming.h [hunk 1]
```diff
@@ -80,7 +80,8 @@ class AP_Arming {
     enum class Required {
         NO           = 0,
         YES_MIN_PWM  = 1,
-        YES_ZERO_PWM = 2
+        YES_ZERO_PWM = 2,
+        MANUAL_PASS_THROUGH = 3
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.h:84]:
```suggestion
        YES_PWM_FROM_RC = 3
```

Comment (tridge) [reply libraries/AP_Arming/AP_Arming.h:84]: i much prefer my name, indicating it is for MANUAL mode, and it doesn't come direct from RC anyway, it just doesn't override to zero

**Comment by tridge** [comment]
> How is this different from arming in manual?

when you arm in manual and get some failsafe types the VTOL motors could start. This is for a 120kg quadplane where you have to stand between the massive VTOL props to start the petrol motor by hand. Being disarmed does give a lot more confidence.

**Comment by tridge** [comment]
> We have some functionality for starting via the ICE library, maybe some extra functionality there would be better?

it is common to need manual throttle control

**Comment by IamPete1** [comment]
> Being disarmed does give a lot more confidence.

But now you have a mode where the forward motor can run when disarmed, thus reducing confidence in the disarmed state. 

In this example the manual start is really just taking the place of the electric starter which we already support. Having manual throttle when using a electric starter would always be useful. We can link the pass through to the ICE engine state. So manual pass-through is only enabled in the starting states but not `ICE_OFF` or `ICE_RUNNING`. We could also add a upper limit param, to constrain the max throttle. So you get manual pass through from 0 to 25% throttle for example, it would limit the potential for accidentally going too high.

Basically I think `ARMING_REQUIRE` is not the right place for this. We could do a better and safer job in the ICE lib. But of course it will cost more flash and maybe need some extra param, but ICE is a compile time option so it it is not so important.

**Comment by tridge** [comment]
> But now you have a mode where the forward motor can run when disarmed, thus reducing confidence in the disarmed state.

it is for petrol motors, without an electric starter. It can't start by itself.

> So manual pass-through is only enabled in the starting states but not ICE_OFF or ICE_RUNNING

you need it for ICE_RUNNING too. You may also need it in ICE_OFF while you are priming the carbie with manual turnover, and you want to do that with ignition off.

File: libraries/AP_ICEngine/AP_ICEngine.cpp [hunk 2]
```diff
@@ -370,6 +373,9 @@ bool AP_ICEngine::throttle_override(float &percentage)
         idle_percent > percentage)
     {
         percentage = idle_percent;
+        if (option_set(Options::THROTTLE_WHILE_DISARMED)) {
```
Comment (IamPete1) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:376]: Don't we need to check if disarmed?

File: libraries/AP_ICEngine/AP_ICEngine.cpp [hunk 3]
```diff
@@ -405,6 +411,12 @@ bool AP_ICEngine::throttle_override(float &percentage)
         return true;
     }
 
+    // if THROTTLE_WHILE_DISARMED is set then we use the base_throttle, allowing the pilot to control throttle while disarmed
+    if (option_set(Options::THROTTLE_WHILE_DISARMED) && base_throttle > percentage) {
```
Comment (IamPete1) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:415]: disarmed check?

File: libraries/AP_ICEngine/AP_ICEngine.cpp [hunk 4]
```diff
@@ -134,8 +134,8 @@ const AP_Param::GroupInfo AP_ICEngine::var_info[] = {
 
     // @Param: OPTIONS
     // @DisplayName: ICE options
-    // @Description: Options for ICE control. The DisableIgnitionRCFailsafe option will cause the ignition to be set off on any R/C failsafe.
-    // @Bitmask: 0:DisableIgnitionRCFailsafe,1:DisableRedineGovernor
+    // @Description: Options for ICE control. The DisableIgnitionRCFailsafe option will cause the ignition to be set off on any R/C failsafe. If ThrottleWhileDisarmed is set then throttle control will be allowed while disarmed for planes when in MANUAL mode.
```
Comment (IamPete1) [inline libraries/AP_ICEngine/AP_ICEngine.cpp:137]: Is it only manual still? Or some others also?

File: ArduPlane/servos.cpp [hunk 5]
```diff
@@ -912,7 +914,7 @@ void Plane::set_servos(void)
     }
 
     float override_pct = SRV_Channels::get_output_scaled(SRV_Channel::k_throttle);
-    if (g2.ice_control.throttle_override(override_pct)) {
+    if (g2.ice_control.throttle_override(override_pct, base_throttle)) {
```
Comment (IamPete1) [inline ArduPlane/servos.cpp:917]: Could use `get_throttle_input(true)` here to always pass manual throttle independent of mode

File: libraries/AP_ICEngine/AP_ICEngine.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:376]: no, percentage==base_throttle when armed.

File: libraries/AP_ICEngine/AP_ICEngine.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:415]: again, don't need to check for how this is used

File: libraries/AP_ICEngine/AP_ICEngine.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:137]: its only manual as the non-manual code paths zero throttle before this point in the plane directory.

File: ArduPlane/servos.cpp (hunk 5 above)
Comment (tridge) [reply ArduPlane/servos.cpp:917]: we don't want it in all modes - imagine we are in AUTO or LOITER or QRTL - we don't want to start passing through throttle when the mode doesn't normally pass thru throttle

File: libraries/AP_ICEngine/AP_ICEngine.cpp (hunk 2 above)
Comment (IamPete1) [reply libraries/AP_ICEngine/AP_ICEngine.cpp:376]: It does at the moment. But not checking leaves a fairly massive bear trap for use to fail into in the future.
